test/helpers/position_assertions.cc

```diff
@@ -851,4 +854,81 @@ string HoverAssertion::toString() const {
+                                                                    string_view assertionType) {
+    static const regex titleVersionRegex(R"(^\[(\w+)\]\s+(.*?)$)");
+
+    smatch matches;
+    string assertionContentsString = string(assertionContents);
+    if (regex_search(assertionContentsString, matches, titleVersionRegex)) {
+        string version = matches[1].str();
+        string title = matches[2].str();
+        return make_shared<ApplyCodeActionAssertion>(filename, range, assertionLine, title, version);
+    }
+
+    ADD_FAILURE_AT(string(filename).c_str(), assertionLine + 1)
+        << fmt::format("Found improperly formatted apply-code-action assertion. Expected apply-code-action "
+                       "[version] code-action-title.");
+    return nullptr;
+}
+ApplyCodeActionAssertion::ApplyCodeActionAssertion(string_view filename, unique_ptr<Range> &range, int assertionLine,
+                                                   string_view title, string_view version)
+    : RangeAssertion(filename, range, assertionLine), title(string(title)), version(string(version)) {}
+
+string ApplyCodeActionAssertion::toString() const {
+    return fmt::format("apply-code-action: [{}] {}", version, title);
+}
+
+void ApplyCodeActionAssertion::check(const UnorderedMap<std::string, std::shared_ptr<core::File>> &sourceFileContents,
+                                     unique_ptr<CodeAction> &codeAction, string_view testName, string_view fileUri) {
+    auto expectedUpdatedFilePath = fmt::format("{}.{}.rbedited", testName, version);
+    string expectedEditedFileContents;
+    try {
+        expectedEditedFileContents = FileOps::read(expectedUpdatedFilePath);
+    } catch (FileNotFoundException e) {
+        ADD_FAILURE() << fmt::format("Missing {} which should contain test file after applying code actions.",
+                                     expectedUpdatedFilePath);
+        return;
+    }
+
+    auto it = sourceFileContents.find(filename);
+    if (it == sourceFileContents.end()) {
+        ADD_FAILURE() << fmt::format("Unable to find referenced source file `{}`", filename);
+        return;
+    }
+
+    auto &file = it->second;
+    for (auto &c : *codeAction->edit.value()->documentChanges) {
+        string actualEditedFileContents = string(file->source());
+        EXPECT_EQ(c->textDocument->uri, fileUri)
+            << fmt::format("Requested code action for {}, but received edits in {}", fileUri, c->textDocument->uri);
+
+        // First, sort the edits by increasing starting location and verify that none overlap.
+        fast_sort(c->edits, [](const auto &l, const auto &r) -> bool { return cmpRanges(*l->range, *r->range); });
+        for (u4 i = 1; i < c->edits.size(); i++) {
+            ASSERT_LT(cmpPositions(*c->edits[i - 1]->range->end, *c->edits[i]->range->start), 0)
+                << fmt::format("Edit {} overlaps edit {}.", c->edits[i - 1]->toJSON(), c->edits[i]->toJSON());
```

Comment:
This error won't give the user any indication of which error has overlapping edits. Maybe change it to something like...?

"Received overlapping autocomplete edits for error `{}`:\n{}\n{}\nThe test runner does not support overlapping autocomplete edits, and it's likely that this is a bug."

Reply:
The corresponding error isn't really available at this level. The edit should have the file though. Tweaked the comment to add more spacing.